File: 772-construct-quad-tree/construct-quad-tree.java
/*
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    
    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
};
*/

class Solution {
    public Node construct(int[][] grid) {
        int n = grid.length;
        return buildQuadTree(grid, 0, 0, n - 1, n - 1);
    }

    private Node buildQuadTree(int[][] grid, int rowStart, int colStart, int rowEnd, int colEnd) {
        if (rowStart == rowEnd && colStart == colEnd) {
            return new Node(grid[rowStart][colStart] == 1, true, null, null, null, null);
        }

        int rowMid = (rowStart + rowEnd) / 2;
        int colMid = (colStart + colEnd) / 2;

        Node topLeft = buildQuadTree(grid, rowStart, colStart, rowMid, colMid);
        Node topRight = buildQuadTree(grid, rowStart, colMid + 1, rowMid, colEnd);
        Node bottomLeft = buildQuadTree(grid, rowMid + 1, colStart, rowEnd, colMid);
        Node bottomRight = buildQuadTree(grid, rowMid + 1, colMid + 1, rowEnd, colEnd);

        if (topLeft.isLeaf && topRight.isLeaf && bottomLeft.isLeaf && bottomRight.isLeaf
            && topLeft.val == topRight.val && topRight.val == bottomLeft.val && bottomLeft.val == bottomRight.val) {
            return new Node(topLeft.val, true, null, null, null, null);
        } else {
            return new Node(true, false, topLeft, topRight, bottomLeft, bottomRight);
        }
    }
}

File: 772-construct-quad-tree/construct-quad-tree.py
"""
# Definition for a QuadTree node.
class Node(object):
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""

class Solution(object):
    def construct(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: Node
        """
        n = len(grid)
        return self.buildQuadTree(grid, 0, 0, n - 1, n - 1)

    def buildQuadTree(self, grid, rowStart, colStart, rowEnd, colEnd):
        if rowStart == rowEnd and colStart == colEnd:
            return Node(grid[rowStart][colStart] == 1, True, None, None, None, None)

        rowMid = (rowStart + rowEnd) // 2
        colMid = (colStart + colEnd) // 2

        topLeft = self.buildQuadTree(grid, rowStart, colStart, rowMid, colMid)
        topRight = self.buildQuadTree(grid, rowStart, colMid + 1, rowMid, colEnd)
        bottomLeft = self.buildQuadTree(grid, rowMid + 1, colStart, rowEnd, colMid)
        bottomRight = self.buildQuadTree(grid, rowMid + 1, colMid + 1, rowEnd, colEnd)

        if topLeft.isLeaf and topRight.isLeaf and bottomLeft.isLeaf and bottomRight.isLeaf and \
                topLeft.val == topRight.val == bottomLeft.val == bottomRight.val:
            return Node(topLeft.val, True, None, None, None, None)
        else:
            return Node(True, False, topLeft, topRight, bottomLeft, bottomRight)
